homeassistant/components/eq3btsmart/__init__.py

```diff
@@ -0,0 +1,146 @@
+from .models import Eq3Config, Eq3ConfigEntryData
+
+PLATFORMS = [
+    Platform.CLIMATE,
+]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Handle config entry setup."""
+
+    mac_address: str = entry.data[CONF_MAC]
+    name: str = entry.data[CONF_NAME]
+
+    eq3_config = Eq3Config(
+        mac_address=mac_address,
+        name=name,
+    )
+
+    device = bluetooth.async_ble_device_from_address(
+        hass, mac_address, connectable=True
+    )
+
+    if device is None:
+        raise ConfigEntryNotReady(f"[{eq3_config.name}] Device could not be found")
+
+    thermostat = Thermostat(
+        thermostat_config=ThermostatConfig(
+            mac_address=mac_address,
+            name=name,
+        ),
+        ble_device=device,
+    )
+
+    eq3_config_entry = Eq3ConfigEntryData(eq3_config=eq3_config, thermostat=thermostat)
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = eq3_config_entry
+
+    entry.async_on_unload(entry.add_update_listener(update_listener))
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    entry.async_create_background_task(
+        hass, _run_thermostat(hass, entry), entry.entry_id
+    )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Handle config entry unload."""
+
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        eq3_config_entry: Eq3ConfigEntryData = hass.data[DOMAIN].pop(entry.entry_id)
+        await eq3_config_entry.thermostat.async_disconnect()
+
+    return unload_ok
+
+
+async def update_listener(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Handle config entry update."""
+
+    await hass.config_entries.async_reload(entry.entry_id)
+
+
+async def _run_thermostat(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Run the thermostat."""
+
+    eq3_config_entry: Eq3ConfigEntryData = hass.data[DOMAIN][entry.entry_id]
+    thermostat = eq3_config_entry.thermostat
+    name = eq3_config_entry.eq3_config.name
+    mac_address = eq3_config_entry.eq3_config.mac_address
+    scan_interval = eq3_config_entry.eq3_config.scan_interval
+
+    await _reconnect_thermostat(hass, entry)
+
+    while True:
+        try:
+            await thermostat.async_get_status()
+        except Eq3Exception as e:
+            if not thermostat.is_connected:
+                _LOGGER.error(
+                    "[%s] eQ-3 device disconnected",
+                    name,
+                )
+                async_dispatcher_send(
+                    hass,
+                    SIGNAL_THERMOSTAT_DISCONNECTED,
+                    mac_address,
+                )
+                await _reconnect_thermostat(hass, entry)
+                continue
+
+            _LOGGER.error(
+                "[%s] Error updating eQ-3 device: %s",
+                name,
+                e,
+            )
```

Comment:
Ideally this only logs once when it fails and does not log again until its failed after it has succeeded so we don't flood the log when its offline

https://developers.home-assistant.io/docs/integration_quality_scale_index?_highlight=logs&_highlight=once#silver-

> Handles device/service unavailable. Log a warning once when unavailable, log once when reconnected.

Reply:
I think this is already the case, since that message only gets logged when an error occurs while the thermostat is connected. If it disconnects, the if-statement above runs and skips the loop iteration after reconnecting.